File: android/src/main/kotlin/video/api/flutter/player/FlutterPlayerController.kt
package video.api.flutter.player

import android.content.Context
import android.util.Log
import android.util.Size
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.view.TextureRegistry
import video.api.player.models.VideoOptions


class FlutterPlayerController(
    private val textureRegistry: TextureRegistry,
    private val messenger: BinaryMessenger,
    private val applicationContext: Context
) : FlutterPlayerInterface {
    private val players = mutableMapOf<Long, FlutterPlayerView>()

    override fun isCreated(textureId: Long): Boolean {
        return players.containsKey(textureId)
    }

    override fun initialize(autoplay: Boolean): Long {
        val player = FlutterPlayerView(
            applicationContext,
            messenger,
            textureRegistry,
            null,
            autoplay
        )
        players[player.textureId] = player

        return player.textureId
    }

    override fun dispose(textureId: Long) {
        players[textureId]?.release() ?: Log.e(TAG, "Unknown player $textureId")
        players.remove(textureId)
    }

    override fun disposeAll() {
        players.values.forEach { it.release() }
        players.clear()
    }

    override fun isPlaying(textureId: Long): Boolean {
        return players[textureId]?.isPlaying ?: run {
            Log.e(TAG, "Unknown player $textureId")
            false
        }
    }

    override fun isLive(textureId: Long): Boolean {
        return players[textureId]?.isLive ?: run {
            Log.e(TAG, "Unknown player $textureId")
            false
        }
    }

    override fun setCurrentTime(textureId: Long, currentTime: Int) {
        players[textureId]?.let { it.currentTime = currentTime.msToFloat() } ?: Log.e(
            TAG,
            "Unknown player $textureId"
        )
    }

    override fun getCurrentTime(textureId: Long): Int {
        return players[textureId]?.currentTime?.toMs() ?: run {
            Log.e(TAG, "Unknown player $textureId")
            0
        }
    }

    override fun getDuration(textureId: Long): Int {
        return players[textureId]?.duration?.toMs() ?: run {
            Log.e(TAG, "Unknown player $textureId")
            0
        }
    }

    override fun getVideoOptions(textureId: Long): VideoOptions? {
        return players[textureId]?.videoOptions ?: run {
            Log.e(TAG, "Unknown player $textureId")
            null
        }
    }

    override fun setVideoOptions(textureId: Long, videoOptions: VideoOptions) {
        players[textureId]?.let { it.videoOptions = videoOptions } ?: Log.e(
            TAG,
            "Unknown player $textureId"
        )
    }

    override fun getAutoplay(textureId: Long): Boolean {
        return players[textureId]?.isAutoplay ?: run {
            Log.e(TAG, "Unknown player $textureId")
            false
        }
    }

    override fun setAutoplay(textureId: Long, autoplay: Boolean) {
        players[textureId]?.let { it.isAutoplay = autoplay } ?: Log.e(
            TAG,
            "Unknown player $textureId"
        )
    }

    override fun getIsMuted(textureId: Long): Boolean {
        return players[textureId]?.isMuted ?: run {
            Log.e(TAG, "Unknown player $textureId")
            false
        }
    }

    override fun setIsMuted(textureId: Long, isMuted: Boolean) {
        players[textureId]?.let { it.isMuted = isMuted } ?: Log.e(
            TAG,
            "Unknown player $textureId"
        )
    }

    override fun getIsLooping(textureId: Long): Boolean {
        return players[textureId]?.isLooping ?: run {
            Log.e(TAG, "Unknown player $textureId")
            false
        }
    }

    override fun setIsLooping(textureId: Long, isLooping: Boolean) {
        players[textureId]?.let { it.isLooping = isLooping } ?: Log.e(
            TAG,
            "Unknown player $textureId"
        )
    }

    override fun getVolume(textureId: Long): Float {
        return players[textureId]?.volume ?: run {
            Log.e(TAG, "Unknown player $textureId")
            0.0F
        }
    }

    override fun setVolume(textureId: Long, volume: Float) {
        players[textureId]?.let { it.volume = volume } ?: Log.e(
            TAG,
            "Unknown player $textureId"
        )
    }

    override fun getVideoSize(textureId: Long): Size? {
        return players[textureId]?.videoSize ?: run {
            Log.e(TAG, "Unknown player $textureId")
            null
        }
    }

    override fun getPlaybackSpeed(textureId: Long): Double {
        return players[textureId]?.playbackSpeed?.toDouble() ?: run {
            Log.e(TAG, "Unknown player $textureId")
            0.0
        }
    }

    override fun setPlaybackSpeed(textureId: Long, playbackSpeed: Double) {
        players[textureId]?.let { it.playbackSpeed = playbackSpeed.toFloat() } ?: Log.e(
            TAG,
            "Unknown player $textureId"
        )
    }

    override fun play(textureId: Long) {
        players[textureId]?.play() ?: Log.e(TAG, "Unknown player $textureId")
    }

    override fun pause(textureId: Long) {
        players[textureId]?.pause() ?: Log.e(TAG, "Unknown player $textureId")
    }

    override fun seek(textureId: Long, offset: Int) {
        players[textureId]?.seek(offset.msToFloat()) ?: Log.e(TAG, "Unknown player $textureId")
    }

    companion object {
        private const val TAG = "FlutterPlayerController"
    }
}

File: ios/Classes/FlutterPlayerController.swift
import ApiVideoPlayer
import Foundation

// Manages a list of multiple players
class FlutterPlayerController {
    private let binaryMessenger: FlutterBinaryMessenger
    private let textureRegistry: FlutterTextureRegistry
    private var players: [Int64: FlutterPlayerView] = [:]

    init(binaryMessenger: FlutterBinaryMessenger, textureRegistry: FlutterTextureRegistry) {
        self.binaryMessenger = binaryMessenger
        self.textureRegistry = textureRegistry
    }

    func isCreated(textureId: Int64) -> Bool {
        return players[textureId] != nil
    }

    func initialize(autoplay: Bool) -> Int64 {
        let player = FlutterPlayerView(binaryMessenger: binaryMessenger, textureRegistry: textureRegistry, autoplay: autoplay)

        players[player.textureId] = player

        return player.textureId
    }

    func dispose(textureId: Int64) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.dispose()
        players.removeValue(forKey: textureId)
    }

    func disposeAll() {
        for player in players.values {
            player.dispose()
        }
        players.removeAll()
    }

    func isPlaying(textureId: Int64) -> Bool {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return false
        }
        return player.isPlaying
    }

    func isLive(textureId: Int64) -> Bool {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return false
        }
        return player.isLive
    }

    func getCurrentTime(textureId: Int64) -> Int {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return 0
        }
        return player.currentTime.toMs()
    }

    func setCurrentTime(textureId: Int64, currentTime: Int) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.currentTime = currentTime.msToCMTime()
    }

    func getDuration(textureId: Int64) -> Int {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return 0
        }
        return player.duration.toMs()
    }

    func getVideoOptions(textureId: Int64) -> VideoOptions? {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return nil
        }
        return player.videoOptions
    }

    func setVideoOptions(textureId: Int64, videoOptions: VideoOptions) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.videoOptions = videoOptions
    }

    func getAutoplay(textureId: Int64) -> Bool {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return false
        }
        return player.autoplay
    }

    func setAutoplay(textureId: Int64, autoplay: Bool) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.autoplay = autoplay
    }

    func getIsMuted(textureId: Int64) -> Bool {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return false
        }
        return player.isMuted
    }

    func setIsMuted(textureId: Int64, isMuted: Bool) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.isMuted = isMuted
    }

    func getIsLooping(textureId: Int64) -> Bool {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return false
        }
        return player.isLooping
    }

    func setIsLooping(textureId: Int64, isLooping: Bool) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.isLooping = isLooping
    }

    func getVolume(textureId: Int64) -> Float {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return 0
        }
        return player.volume
    }

    func setVolume(textureId: Int64, volume: Float) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.volume = volume
    }

    func getVideoSize(textureId: Int64) -> CGSize? {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return nil
        }
        return player.videoSize
    }

    func play(textureId: Int64) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.play()
    }

    func pause(textureId: Int64) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.pause()
    }

    func seek(textureId: Int64, offset: Int) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.seek(offset: offset.msToCMTime())
    }

    func setPlaybackSpeed(textureId: Int64, speedRate: Double) {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return
        }
        player.speedRate = Float(speedRate)
    }

    func getPlaybackSpeed(textureId: Int64) -> Double {
        guard let player = players[textureId] else {
            print("Unknown player \(textureId)")
            return 0
        }
        return Double(player.speedRate)
    }
}

extension Int {
    func msToCMTime() -> CMTime {
        return CMTime(value: CMTimeValue(self), timescale: 1000)
    }
}

extension CMTime {
    func toMs() -> Int {
        let seconds = self.seconds
        guard !(seconds.isNaN || seconds.isInfinite) else {
            return 0
        }
        return Int(seconds * 1000)
    }
}
